File: ML.py
import pandas as pd;
df = pd.read_csv('merged_data.csv')

def match_courses_with_preferences(preferences, roll_no, df):
    # Replace NaN values in 'Keywords' column with empty strings
    df['Keywords'] = df['Keywords'].fillna('')

    if preferences:
        # Filter courses based on matching keywords
        matched_courses = df[df['Keywords'].str.contains(preferences, case=False)]
    else:
        # Extract the third and fourth letters from the roll number
        code = roll_no[2:4]
        # Filter courses based on matching the extracted code with the course code
        matched_courses = df[df['Course code'].str.startswith(code)]

    # Sort courses based on the number of matched keywords
    matched_courses['Matched Keywords'] = matched_courses['Keywords'].apply(
        lambda x: sum(1 for word in x.split() if word.lower() in preferences.lower()))
    
    # Sort matched courses based on the number of matched keywords
    matched_courses = matched_courses.sort_values(by='Matched Keywords', ascending=False)

    # Return course codes of matched courses
    return matched_courses['Course code'].tolist()

# Example usage
preferences = ''  # Set preferences here
roll_no = '22CY10043'  # Set roll number here
course_codes = match_courses_with_preferences(preferences, roll_no, df)
print(course_codes)

File: ML.js
const fs = require('fs');
const csv = require('csv-parser');

const match_courses_with_preferences = (preferences, roll_no, data) => {
    data.forEach(row => {
        row['Keywords'] = row['Keywords'] || '';
    });

    let matched_courses;
    if (preferences) {
        matched_courses = data.filter(row => row['Keywords'].toLowerCase().includes(preferences.toLowerCase()));
    } else {
        const code = roll_no.slice(2, 4);
        matched_courses = data.filter(row => row['Course code'].startsWith(code));
    }

    matched_courses.forEach(row => {
        row['Matched Keywords'] = row['Keywords'].split(' ').filter(word => preferences.toLowerCase().includes(word.toLowerCase())).length;
    });

    matched_courses.sort((a, b) => b['Matched Keywords'] - a['Matched Keywords']);

    return matched_courses.map(row => row['Course code']);
};

const data = [];
fs.createReadStream('merged_data.csv')
    .pipe(csv())
    .on('data', row => {
        data.push(row);
    })
    .on('end', () => {
        const preferences = 'tech';
        const roll_no = '22CY10043';
        const course_codes = match_courses_with_preferences(preferences, roll_no, data);
        console.log(course_codes);
    });